File: screenshot.cs
// screenshot.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.Windows.Forms;
using System.Threading;

class Screenshot
{
    static void Main(string[] args)
    {
        string output = null;
        int delay = 0;
        string region = null;
        for (int i = 0; i < args.Length; i++)
        {
            if (args[i] == "-o" && i+1 < args.Length)
                output = args[++i];
            else if (args[i] == "-d" && i+1 < args.Length)
                delay = int.Parse(args[++i]);
            else if (args[i] == "-r" && i+1 < args.Length)
                region = args[++i];
            else if (args[i] == "-h" || args[i] == "--help")
            {
                Console.WriteLine("Usage: screenshot [options]");
                Console.WriteLine("  -o <file>   Output file");
                Console.WriteLine("  -d <sec>    Delay in seconds");
                Console.WriteLine("  -r <x,y,w,h> Region");
                Console.WriteLine("  -h          Help");
                return;
            }
        }
        if (delay > 0)
        {
            Console.WriteLine($"Ожидание {delay} секунд...");
            Thread.Sleep(delay * 1000);
        }
        if (string.IsNullOrEmpty(output))
            output = $"screenshot_{DateTime.Now:yyyyMMdd_HHmmss}.png";

        try
        {
            Bitmap bitmap;
            if (!string.IsNullOrEmpty(region))
            {
                var parts = region.Split(',');
                if (parts.Length != 4)
                {
                    Console.WriteLine("Неверный формат области. Используйте x,y,w,h");
                    return;
                }
                int x = int.Parse(parts[0]);
                int y = int.Parse(parts[1]);
                int w = int.Parse(parts[2]);
                int h = int.Parse(parts[3]);
                bitmap = new Bitmap(w, h);
                using (Graphics g = Graphics.FromImage(bitmap))
                {
                    g.CopyFromScreen(x, y, 0, 0, new Size(w, h));
                }
            }
            else
            {
                var screen = Screen.PrimaryScreen.Bounds;
                bitmap = new Bitmap(screen.Width, screen.Height);
                using (Graphics g = Graphics.FromImage(bitmap))
                {
                    g.CopyFromScreen(0, 0, 0, 0, screen.Size);
                }
            }
            bitmap.Save(output, ImageFormat.Png);
            Console.WriteLine($"Скриншот сохранён в {output}");
        }
        catch (Exception e)
        {
            Console.WriteLine($"Ошибка: {e.Message}");
        }
    }
}

File: screenshot.rb
#!/usr/bin/env ruby
# screenshot.rb
# encoding: UTF-8

require 'rmagick'
include Magick
require 'optparse'
require 'time'

options = { output: nil, delay: 0, region: nil }

OptionParser.new do |opts|
  opts.banner = "Usage: screenshot.rb [options]"
  opts.on("-o", "--output FILE", "Выходной файл") { |v| options[:output] = v }
  opts.on("-d", "--delay SEC", Integer, "Задержка в секундах") { |v| options[:delay] = v }
  opts.on("-r", "--region X,Y,W,H", String, "Область захвата") { |v| options[:region] = v }
  opts.on("-h", "--help", "Справка") { puts opts; exit }
end.parse!

if options[:delay] > 0
  puts "Ожидание #{options[:delay]} секунд..."
  sleep options[:delay]
end

output = options[:output] || "screenshot_#{Time.now.strftime('%Y%m%d_%H%M%S')}.png"

begin
  if options[:region]
    x, y, w, h = options[:region].split(',').map(&:to_i)
    # RMagick не имеет прямого захвата экрана, используем системную команду? 
    # Вместо этого используем системную утилиту (как в C++).
    # Для простоты вызовем внешнюю утилиту.
    # На Linux используем import, на macOS screencapture, на Windows PowerShell.
    if RUBY_PLATFORM =~ /linux/
      system("import -window root -crop #{w}x#{h}+#{x}+#{y} #{output}")
    elsif RUBY_PLATFORM =~ /darwin/
      system("screencapture -R #{x},#{y},#{w},#{h} #{output}")
    elsif RUBY_PLATFORM =~ /mswin|mingw|cygwin/
      # PowerShell
      system("powershell -command \"Add-Type -AssemblyName System.Drawing; $b=New-Object System.Drawing.Bitmap(#{w},#{h}); $g=[System.Drawing.Graphics]::FromImage($b); $g.CopyFromScreen(#{x},#{y},0,0,New-Object System.Drawing.Size(#{w},#{h})); $b.Save('#{output}','png')\"")
    end
  else
    # Захват всего экрана через системные утилиты
    if RUBY_PLATFORM =~ /linux/
      system("import -window root #{output}")
    elsif RUBY_PLATFORM =~ /darwin/
      system("screencapture -x #{output}")
    elsif RUBY_PLATFORM =~ /mswin|mingw|cygwin/
      system("powershell -command \"Add-Type -AssemblyName System.Drawing; $b=New-Object System.Drawing.Bitmap([System.Windows.Forms.Screen]::PrimaryScreen.Bounds.Width, [System.Windows.Forms.Screen]::PrimaryScreen.Bounds.Height); $g=[System.Drawing.Graphics]::FromImage($b); $g.CopyFromScreen(0,0,0,0,$b.Size); $b.Save('#{output}','png')\"")
    end
  end
  puts "Скриншот сохранён в #{output}"
rescue => e
  puts "Ошибка: #{e.message}"
  exit 1
end
